### PR TITLE
Mysql.pm - Check CSF Ports

### PR DESCRIPTION
**Pull request by sporks5000**
Previously this module only checked for iptables rules that would deny the
mysql port, however by default csf denies all ports, and then allows only
those specified in the configuration, so the rules it's looking for would not
be present. In addition to the check against iptables, I added a check to see
if the csf.conf file exists, and whether or not the mysql port is open. Also,
I modified the suggestions to recommend either changing the bind address OR
closing the port.

### CODE REVIEW

File: pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm [hunk 1]
```diff
@@ -165,25 +165,41 @@ sub _check_for_public_bind_address {
...
+                    $port
+                ],
             );
         }
     }
 
     return 1;
 }
 
+sub csf_port_closed {
```
Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: According to [0], ports can be listed with ranges, which this code does not account for.

Consider something like:

```perl
sub config_key_contains_port {
    my ( $file, $key, $port ) = @_;

    my $csf_conf = Cpanel::LoadFile::load_if_exists($file);
    return if !$csf_conf;

    foreach my $line ( split m/\n/, $csf_conf ) {
        if ( $line =~ m/^\s*\Q$key\E\s*=\s*(['"])([^'"]*)\1/a ) {
            my $port_list = $2;
            foreach my $entry ( split m/,/, $port_list ) {
                my ( $first, $last ) = split m/:/, $entry;
                if ($last) {
                    return 1 if $first <= $port && $port <= $last;
                }
                else {
                    return 1 if $port == $first;
                }
            }
        }
    }

    return 0;
}

sub csf_port_closed {
    my ($port) = @_;
    return config_key_contains_port( '/etc/csf/csf.conf', 'TCP_IN', $port );
}
```

[0] https://github.com/omega8cc/boa/blob/2e6b4ec963f3e5da0a1693648c8514eedda015ba/aegir/conf/csf.conf#L111

**Comment by ColMelvin** [comment]
Please add `use Cpanel::LoadFile ();`.

File: pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm [hunk 2]
```diff
@@ -165,25 +165,41 @@ sub _check_for_public_bind_address {
         else {
```
Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: I suspect you also want to add a clause for `elsif ( csf_port_closed($port) )` here too.

Comment (sporks5000) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: Can you think of any reason against just tacking on `|| ( csf_port_closed($port) )` near the end of line 162 like this:

```
elsif ( ( ( $version == 4 ) && @deny_rules && ( ( $bind_address =~ /ffff/i ) ? @deny_rules_6 : 1 ) ) || ( ( $version == 6 ) && @deny_rules_6 ) || ( csf_port_closed($port) ) ) {
```

Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: That is just fine too.  GitHub wouldn't let me comment on line 162, so I put it here instead.

File: pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm (hunk 1 above)
Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: Oops, that should be negated: `return !config_key_contains_port( '/etc/csf/csf.conf', 'TCP_IN', $port );`.

Comment (sporks5000) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: I feel that it would be morally wrong for me to just take what you've written above and throw it in with my pull request, but at the same time, I cannot think of anything that I would change about it. Would it be acceptable for me to just finalize the other changes that you've recommended and have you make that change as a separate commit?

Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: Actually, I found another error.  If the file doesn't exist, the port is said to be blocked.  The code should be:

```perl
sub csf_port_closed {
    my ($port) = @_;
    my $contains = config_key_contains_port( '/etc/csf/csf.conf', 'TCP_IN', $port );
    return if !defined $contains;
    return !$contains;
}
```

That said, you are welcome to copy&paste my code snippets into code that you are submitting to this project; no need for attribution (after all, the pull request has this information anyway).

Comment (sporks5000) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: That's fair. Thank you - I have made the recommended changes.

**Comment by ColMelvin** [comment]
@bk2204 Can you review?  I contributed to the changes.

**Comment by bk2204** [comment]
Of course.

File: pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm [hunk 3]
```diff
@@ -159,31 +160,68 @@ sub _check_for_public_bind_address {
...
 }
 
+sub config_key_contains_port {
+    my ( $file, $key, $port ) = @_;
+
+    my $csf_conf = Cpanel::LoadFile::load_if_exists($file);
+    return if !$csf_conf;
+
+    foreach my $line ( split m/\n/, $csf_conf ) {
+        if ( $line =~ m/^\s*\Q$key\E\s*=\s*(['"])([^'"]*)\1/a ) {
```
Comment (bk2204) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: I like the use of \1 here to match the quotes properly.

**Comment by bk2204** [comment]
Looks good to me.

File: pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm (hunk 3 above)
Comment (sporks5000) [inline pkg/Cpanel/Security/Advisor/Assessors/Mysql.pm]: Yeah, @ColMelvin is kind of amazing.
